csrc/scheduler/transpose.cpp

```diff
@@ -714,29 +714,138 @@ std::shared_ptr<TransposeParams> getTransposeHeuristics(
       ceilDiv((int64_t)params->tile_size1 * (int64_t)params->tile_size2, 32l);
   max_unroll_factor = std::min(max_unroll_factor, max_unroll_factor_block);
 
-  // Compute maximum vectorize factor that can be used
-  size_t vectorize_factor1 = max_unroll_factor;
-  size_t vectorize_factor2 = max_unroll_factor;
-
-  for (auto tv : grouped_inputs_outputs[0]) {
-    const auto tv_vectorize_factor =
-        runtime_info.getMaxVectorizableWidth(tv, false);
-    vectorize_factor1 = std::min(vectorize_factor1, tv_vectorize_factor);
-  }
-  // TODO: Since group2 only has global->shared and shared->global set op, we
-  // can have fine-grained control of unroll/vectorization at per tensor level.
-  // We should not be using a single global vectorize factor for the entire
-  // group 2
-  for (auto tv : grouped_inputs_outputs[1]) {
-    const auto tv_vectorize_factor =
-        runtime_info.getMaxVectorizableWidth(tv, false);
-    vectorize_factor2 = std::min(vectorize_factor2, tv_vectorize_factor);
-  }
+  // Note: [Computing Vectorization Width for Transpose]
+  //
+  // With support of small transpose dimension (see Note [Supporting small
+  // transpose dimensions]), we need to consider the transformation applied on
+  // our tile to compute the safe vectorization width. e.g. For a simple
+  // transpose:
+  //    (i0, i1, i2, i3, i4) size (2, 3, 65536, 4, 7)
+  // -> (i3, i4, i2, i0, i1)
+  //
+  // transpose scheduler would apply transformation and tile on virtual
+  // innermost dimensions. ( (i0*i1*i2/2), (2*i3*i4)/32) So we need to use the
+  // size of the virtual innermost dimensions to compute our vectorization
+  // width. In the example above, we are looking at 2*3*65536/2, 2*4*7.
+  //
+  // Currently there's limitation on our iter domain mapping. Since we can only
+  // do it on rfactor/root domain, we cannot map across `split` domains. So the
+  // example above will only have vectorization size of 2 and 4 repsectively for
+  // the merge virtual innermost dimensions, rather than considering the split
+  // and merged i2/2 & 2.
+  //
+  // TODO: We use ContiguousInnerDimensionsMapper to compute the size of virtual
+  // innermost dimension. The analysis right now is limited on rfactor domain
+  // only, so we can't actually map the `split` iter domains, which limits the
+  // vectorization width we can apply. We need to fix that.
```

Comment:
I think the problem is bigger than that. Because the vectorization factor can be incompatible with tiling. For example, what if the tile size is `7`? (Could you add a test for it? Also, it worth testing the case where the virtual inner dim is 1024*7 but the tile size is 2*7). I think the TODO item for starting propagating from leaf instead of rFactor domain is required, not optional or future work, because otherwise we can not correctly handle tiling? Also, it would make the analysis here much easier: we can factor out part of `scheduleTranspose` out, and just apply it here on `cloned_1_td` and begin the analysis on leaf. If so, probably we should do it in a separate PR again, and test it separately?  There are a lot of vectorization helper tests in https://github.com/NVIDIA/Fuser/blob/main/test/test_gpu_utils.cpp

Reply:
I'll try to add a test for odd tile sizes. I'm not quite sure I follow how `odd` tile size would break it here. Doesn't an odd tile size mean any inner dimension mapping would be `odd` and we wouldn't have any vectorization logic enabled right now?

Other than that, all parts regarding refactor and simplify `scheduleTranspose` sounds good to me here.